### PR TITLE
CodeQL migration: Java topics - change titles & add intros (2164)

### PR DESCRIPTION
**Pull request by hubwriter**
product-documentation issue [#2164](https://github.com/github/product-documentation/issues/2164)

### Summary

This changes the titles of the Java topics in the "Learning" section, and adds intros as the first paragraph. 

I've also made some small editorial changes to bring the style a little more in line with GitHub's writing style. However, I haven't made any larger content changes or structural changes.

### Changes in this PR

| Article preview | Source | Notes |
| --- | --- | --- |
"[CodeQL for Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/ql-for-java.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-f7a24c1793992f4b1b91c790c0f9a3c6) | 
"[CodeQL library for Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/introduce-libraries-java.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-7666fe4088a3b803151f2b7388bd18bf) | 
"[Analyzing data flow in Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/dataflow.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-625b2b40b12b42abfc794b204b89704b) | 
"[Types in Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/types-class-hierarchy.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-2624db6ce320b365148a71c2af656589) | I think "Java types" would be a better title. "Types in Java" sounds like this article is explaining types in Java, when in fact it is explaining how to use CodeQL to write queries that investigate the use of particular data types in Java code.
"[Expressions and statements in Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/expressions-statements.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-d40bbaa818d7c9cbaa54bee58ce53983) | The title is not good. This topic is about writing queries to identify comparisons between different data types where one side of the comparison might overflow. The title is misleadingly general. It's quite a specific little topic. I'd suggest something like: Overflow-prone comparisons in Java
"[Navigating the call graph](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/call-graph.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-50c96fa4bdf923a8a03d98801bbef489) | 
"[Annotations in Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/annotations.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-90ae86b5c1efc71d4498670334b433d8) | 
"[Javadoc](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/javadoc.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-4902835928c29eb3eb58b8dc44092af6) | 
"[Working with source locations](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/source-locations.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-0738a1d3a4cdfb0e97243bccc35b18c6) | 
"[Abstract syntax tree classes in Java](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/ast-class-reference.html)" | [Source](https://github.com/Semmle/ql/pull/2866/files#diff-c2659f7d607d074394f7bb8d1d606c89) | I think the inclusion of "Abstract syntax tree" in this heading is unnecessary. This reference topic does not mention the AST anywhere. I think "CodeQL classes for Java" might be a better title. I'm assuming that these classes *are* CodeQL classes and not Java classes (which "classes in Java" leads you to think). Alternatively, add some content near the start of the topic mentioning the AST and its nodes.

### Points needing attention

- Some topics contain links to queries in the query console of LGTM.com. This is done using the linked text "See this in the query console." Reading this as a newcomer this seems strange with no context, and it's going to seem stranger once this is in GitHub documentation as it takes you to a website that readers may not have heard of and, currently at least, seems to have no connection with GitHub. If these links are to stay then I think the text needs to say "See this in the query console on LGTM.com." 
- I've changed "What next?" at the end of some topics to "Further reading". However, I haven't altered the content of these sections, which is discursive, rather than the GitHub style of just having links (one linked article per bullet point). These probably need redone in GitHub style, but some probably need pruned down to avoid too long a bullet list.
- I've removed a few instances of "thus", which is overly stuffy/formal (if not archaic) and doesn't belong in software documentation. We should remove this word throughout (it only appears in one existing GitHub article).

### CODE REVIEW

**Comment by felicitymay** [comment]
> Some topics contain links to queries in the query console of LGTM.com. This is done using the linked text "See this in the query console." Reading this as a newcomer this seems strange with no context, and it's going to seem stranger once this is in GitHub documentation as it takes you to a website that readers may not have heard of and, currently at least, seems to have no connection with GitHub. If these links are to stay then I think the text needs to say "See this in the query console on LGTM.com."

That's a good point. When the link text was written, this documentation was embedded in LGTM. Clearly we should have thought to update the text when we moved the documentation out of LGTM. Given that this is used across the entire set of documents, please can you raise an issue. It sounds like we need to decide whether to keep these links and, if so, what the new link text should be.

**Comment by felicitymay** [comment]
Your comments on the "Types in Java" and "Expressions and statements in Java" titles - this is a good point and probably also applies to similar topics/titles in Python. Possibly, once we've discussed the content type of these topics with Emily, it will be easier to define a more descriptive title.

**Comment by felicitymay** [comment]
> Title: Abstract syntax tree classes in Java

What you say about this topic makes sense. Based on the information in the library overview section "[Abstract syntax tree](http://docteam.internal.semmle.com/alistair/CodeQL/migration-prep/pd-issue-2164/docs/language/learn-ql/_build/java/introduce-libraries-java.html#abstract-syntax-tree)", it sounds as if your second suggestion might be the right answer. That is, add a short conceptual section explaining that these are the CodeQL classes used to model Java statements and expressions.

File: docs/language/learn-ql/java/source-locations.rst [hunk 1]
```diff
@@ -1,10 +1,12 @@
-Tutorial: Working with source locations
-=======================================
+Working with source locations
+=============================
 
-Overview
---------
+You can use the location of entities within Java code to look for potential errors. Locations allow you to deduce the presence, or absence, of white space which, in some cases, may indicate a problem.
```
Comment (felicitymay) [inline docs/language/learn-ql/java/source-locations.rst]: The second sentence could perhaps be simplified?

```suggestion
You can use the location of entities within Java code to look for potential errors. Locations allow you to deduce how white space is used which, in some cases, may indicate a problem. 
```

**Comment by hubwriter** [comment]
Created an issue to discuss/decide about text of links to the query console:
https://github.com/github/product-documentation/issues/2237

File: docs/language/learn-ql/java/source-locations.rst (hunk 1 above)
Comment (hubwriter) [reply docs/language/learn-ql/java/source-locations.rst]: I think the longer version is clearer.

File: docs/language/learn-ql/java/call-graph.rst [hunk 2]
```diff
@@ -1,5 +1,7 @@
-Tutorial: Navigating the call graph
-===================================
+Navigating the call graph
+=========================
+
+CodeQL provides an API for identifying code that calls other code, and code that can be called from elsewhere. This allows you to find, for example, methods that are never used.
```
[the remark this answers is not in the document]
Comment (hubwriter) [reply docs/language/learn-ql/java/call-graph.rst]: done

**Comment by jf205** [comment]
@hubwriter this looks pretty good to me. I'd be in favour of merging this (pending approval from @shati-patel and @felicitymay), with a view to having another pass through when the other 'CodeQL for X' PRs have all been completed. What do you think?

**Comment by hubwriter** [comment]
@jf205 - thanks, I'm happy to merge this whenever.
@felicitymay was worried that there would be problems because I merged master into this branch, so I've held off merging. I can easily remove that merge though.

**Comment by jf205** [comment]
> @felicitymay was worried that there would be problems because I merged master into this branch

Fair point. Looks like you could just drop that commit before you merge.

**Comment by jf205** [comment]
@hubwriter: https://github.com/Semmle/ql/pull/3035 is now merged, so it looks like you can cleanly merge this PR without dropping `Merge branch 'master' into alistairs-docs-preparation-1` (fe9f9743531005af33a504e775b9962a5a97b943)

**Comment by hubwriter** [comment]
Merging

**Pull request merged**